cmake/cpu_info.cmake

```diff
@@ -15,25 +15,51 @@
 
 # Symbols with information about the CPU.
 
-IF(CMAKE_SYSTEM_NAME MATCHES "Darwin")
-  FIND_PROGRAM(SYSCTL sysctl)
-  MARK_AS_ADVANCED(SYSCTL)
+IF(NOT CPU_LEVEL1_DCACHE_LINESIZE)
+  IF(CMAKE_CROSSCOMPILING)
+   MESSAGE(FATAL_ERROR
+   "CPU_LEVEL1_DCACHE_LINESIZE is not defined.  Please specify -DCPU_LEVEL1_DCACHE_LINESIZE "
+   "to the size of the level 1 data cache for the destination architecture.")
+  ELSE()
+    IF(CMAKE_SYSTEM_NAME MATCHES "Darwin")
+      FIND_PROGRAM(SYSCTL sysctl)
+      MARK_AS_ADVANCED(SYSCTL)
 
-  IF(SYSCTL)
-    EXECUTE_PROCESS(
-      COMMAND ${SYSCTL} -n hw.cachelinesize
-      OUTPUT_VARIABLE CPU_LEVEL1_DCACHE_LINESIZE
-      )
-  ENDIF()
+      IF(SYSCTL)
+        EXECUTE_PROCESS(
+          COMMAND ${SYSCTL} -n hw.cachelinesize
+          OUTPUT_VARIABLE CPU_LEVEL1_DCACHE_LINESIZE
+          OUTPUT_STRIP_TRAILING_WHITESPACE
+          )
+      ENDIF()
+
+    ELSE()
+      FIND_PROGRAM(GETCONF getconf)
+      MARK_AS_ADVANCED(GETCONF)
 
-ELSE()
-  FIND_PROGRAM(GETCONF getconf)
-  MARK_AS_ADVANCED(GETCONF)
+      IF(GETCONF)
+        EXECUTE_PROCESS(
+          COMMAND ${GETCONF} LEVEL1_DCACHE_LINESIZE
+          OUTPUT_VARIABLE CPU_LEVEL1_DCACHE_LINESIZE
+          OUTPUT_STRIP_TRAILING_WHITESPACE
+          )
+      ENDIF()
+    ENDIF()
+  ENDIF()
 
-  IF(GETCONF)
-    EXECUTE_PROCESS(
-      COMMAND ${GETCONF} LEVEL1_DCACHE_LINESIZE
-      OUTPUT_VARIABLE CPU_LEVEL1_DCACHE_LINESIZE
-      )
+  IF(CPU_LEVEL1_DCACHE_LINESIZE)
+    SET(CPU_LEVEL1_DCACHE_LINESIZE "${CPU_LEVEL1_DCACHE_LINESIZE}" CACHE INTERNAL "CPU level 1 Data cache line size")
```

Comment:
I'd strongly prefer to have just one ```SET(...CACHE...)```. In other words move this ```SET()``` somewhere after last ENDIF() and make other ```SET()'s``` not cached.

Also CMake suggests that we don't mix cached and not cached variables with the same name. I'd probably do something like ```SET(TMP 64)``` in the beginning, then use ```TMP``` as a destination for ```EXECUTE_PROCESS()``` etc, then ```SET(...CACHE...)``` then ```UNSET(TMP)```.